### PR TITLE
Wired up the optionDAO

### PR DESCRIPTION
**Pull request by tharmaman**

### CODE REVIEW

File: src/foam/dao/OptionDAO.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/dao/OptionDAO.js]: Is this new code?

File: src/foam/dao/OptionDAO.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      documentation: 'OptionDAO.optionPredicate is checked against an object; if returns true, OptionDAO.optionDelegate is used.'+
+      'Default uses OptionDAO.delegate.'
+    },
+  ],
+
+  methods: [
+    {
+      name: 'put_',
+      javaCode: `
+        if  ( getOptionPredicate().f(obj) ) {
```
Comment (kgrgreer) [inline src/foam/dao/OptionDAO.js]: A bit nicer way to do this with less repetition would be to have a method named something like getDelegateFor(FObject obj), then this method would be: return getDelegateFor(obj).put_(x, obj);
Similar for remove_() below. Then the dao selection logic isn't duplicated.

File: src/foam/dao/OptionDAO.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'OptionDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: `
```
Comment (kgrgreer) [inline src/foam/dao/OptionDAO.js]: Please improve/expand upon the documentation. I couldn't understand what you meant from reading the documentation and had to read the code figure things out.

File: src/foam/dao/OptionDAO.js (hunk 1 above)
Comment (tharmaman) [reply src/foam/dao/OptionDAO.js]: Moveed to 2020.

File: src/foam/dao/OptionDAO.js (hunk 2 above)
Comment (tharmaman) [reply src/foam/dao/OptionDAO.js]: Added.

File: src/foam/dao/OptionDAO.js (hunk 3 above)
Comment (tharmaman) [reply src/foam/dao/OptionDAO.js]: Ya added more documentation.

File: src/foam/dao/OptionDAO.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    OptionDAO allows the user to pass a delegate to either the default delegate (OptionDAO.delegate) or
```
Comment (jlhughes) [inline src/foam/dao/OptionDAO.js]: I find you documentation a little unclear.  How about: 

This DAO supports branching to another DAO when the predicate evaluates to true.

File: src/foam/dao/OptionDAO.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'OptionDAO',
```
Comment (kgrgreer) [inline src/foam/dao/OptionDAO.js]: I'm not too keen on this name. How about something like PredicatedDualDelegateDAO, which provides more information than "Option", whose meaning isn't meaningful?

Comment (jlhughes) [reply src/foam/dao/OptionDAO.js]: I was thinking BranchingDAO

Comment (nick-prat) [reply src/foam/dao/OptionDAO.js]: Maybe SwitchDAO?

Comment (jlhughes) [reply src/foam/dao/OptionDAO.js]: @tharmaman ok, "we've" decided on Kevin's name PredicatedDualDelegateDAO

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.dao.NullDAO'
+  ],
+
+  properties: [
+    {
+      class: 'Proxy',
+      of: 'foam.dao.DAO',
+      name: 'optionDelegate',
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
      name: 'predicatedDelegate',
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      swiftPostSet: `
+        if let oldValue = oldValue as? foam_dao_AbstractDAO {
+          _ = oldValue.on["reset"].pub()
+        }
+      `
+      ,
+    },
+    {
+      class: 'foam.mlang.predicate.PredicateProperty',
+      name: 'optionPredicate',
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
      name: 'predicate',
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      factory: function () {
+        return foam.mlang.predicate.True.create();
+      },
+      javaFactory: `
+      return foam.mlang.MLang.TRUE;
+      `,
+      documentation: `
+        PredicatedDualDelegateDAO.optionPredicate is checked against an object;
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
        PredicatedDualDelegateDAO.predicate is checked against an object; 
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      name: 'optionPredicate',
+      factory: function () {
+        return foam.mlang.predicate.True.create();
+      },
+      javaFactory: `
+      return foam.mlang.MLang.TRUE;
+      `,
+      documentation: `
+        PredicatedDualDelegateDAO.optionPredicate is checked against an object; 
+        if returns true, PredicatedDualDelegateDAO.optionDelegate is used.
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
        if returns true, PredicatedDualDelegateDAO.predicatedDelegate is used.
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      `
+    },
+    {
+      name: 'getDelegateFor',
+      type: 'foam.dao.DAO',
+      args: [
+        { name: 'obj', type: 'foam.core.FObject' },
+      ],
+      javaCode: `
+        if ( getOptionPredicate().f(obj) ) {
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
        if ( getPredicate().f(obj) ) {
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+          return getOptionDelegate();
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
          return getPredicatedDelegate();
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'PredicatedDualDelegateDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: `
+    PredicatedDualDelegateDAO allows the user to pass a delegate to either 
+    the default delegate (PredicatedDualDelegateDAO.delegate) or
+    the option Delegate (PredicatedDualDelegateDAO.optionDelegate).
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
    the option Delegate (PredicatedDualDelegateDAO.predicatedDelegate). 
```

File: src/foam/dao/PredicatedDualDelegateDAO.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  package: 'foam.dao',
+  name: 'PredicatedDualDelegateDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: `
+    PredicatedDualDelegateDAO allows the user to pass a delegate to either 
+    the default delegate (PredicatedDualDelegateDAO.delegate) or
+    the option Delegate (PredicatedDualDelegateDAO.optionDelegate). 
+    Passing to the option delegate will occur if the
+    PredicatedDualDelegateDAO.optionPredicate evaluates to true.
```
Comment (jlhughes) [inline src/foam/dao/PredicatedDualDelegateDAO.js]:
```suggestion
    PredicatedDualDelegateDAO.predicate evaluates to true.
```

**Pull request merged**